src/Views/OperatingSystemView.vala

```diff
@@ -222,10 +319,123 @@ public class About.OperatingSystemView : Gtk.Box {
+
+                try {
+                    var details = yield update_proxy.get_update_details ();
+                    updates_description.label = ngettext (
+                        "%i update available",
+                        "%i updates available",
+                        details.packages.length
+                    ).printf (details.packages.length);
+
+                    packages.splice (0, packages.get_n_items (), details.packages);
+                } catch (Error e) {
+                    updates_description.label = _("Unable to determine number of updates");
+                    warning ("Failed to get updates list from backend: %s", e.message);
+                }
+                break;
+            case DOWNLOADING:
+                updates_image.icon_name = "browser-download";
+                updates_title.label = _("Downloading Updates");
+                updates_description.label = current_state.message;
+                button_stack.visible_child_name = "cancel";
+                break;
+            case RESTART_REQUIRED:
+                updates_image.icon_name = "system-reboot";
+                updates_title.label = _("Restart Required");
+                updates_description.label = _("A restart is required to finish installing updates");
+                button_stack.visible_child_name = "blank";
+                break;
+            case ERROR:
+                updates_image.icon_name = "dialog-error";
+                updates_title.label = _("Failed to download updates");
+                updates_description.label = _("Manually refreshing updates may resolve the issue.");
+                button_stack.visible_child_name = "refresh";
+                break;
+        }
+    }
+
+    private void details_clicked () {
+        if (current_state == null) {
+            return;
+        }
+
+        if (current_state.state == ERROR) {
+            var message_dialog = new Granite.MessageDialog (
+                _("Failed to download updates"),
+                _("This may have been caused by sideloaded or manually compiled software, a third-party software source, or a package manager error. Manually refreshing updates may resolve the issue."),
+                new ThemedIcon ("dialog-error")
+            ) {
+                transient_for = (Gtk.Window) get_root (),
+                modal = true
+            };
+
+            message_dialog.show_error_details (current_state.message);
+
+            message_dialog.response.connect (message_dialog.destroy);
+            message_dialog.present ();
+            return;
+        }
+
+        var details_dialog = new UpdateDetailsDialog (packages) {
+            transient_for = (Gtk.Window) get_root ()
+        };
+        details_dialog.present ();
+    }
+
+    private async void refresh_clicked () {
+        if (update_proxy == null) {
+            return;
+        }
+
+        try {
+            yield update_proxy.check_for_updates (current_state.state == ERROR, false);
```

Comment:
I'm not sure about this "force" behavior. This means that the refresh button doesn't have any effect sometimes. I think if someone requests a refresh we should always honor that

Reply:
Hmm yeah I guess that makes sense. You can push if you want, I'm currently not at my computer :(

Reply:
Done